File: packages/manifest/lib/src/extractors/aspects.dart
import '../raw/raw_types.dart';
import '../types/records.dart';
import '../types/services.dart';
import '../types/stores.dart';
import 'common.dart';

final _aspectCategoryRe = RegExp(
  r'^(titan|hunter|warlock|shared)\.(arc|solar|void|stasis|strand|prismatic)\.aspects$',
);
const _aspectEnergyStatHash = 2223994109;

class AspectsExtractor implements EntityExtractor {
  @override
  MvpStoreName get store => MvpStoreName.aspects;

  @override
  Future<List<Object>> extract(LoadRawTable loadTable) async {
    final itemTable = await loadTable('DestinyInventoryItemDefinition');
    final result = <AspectRecord>[];

    for (final item in iterItems(itemTable)) {
      if (!_isAspectItem(item)) continue;
      final typeName = item.itemTypeDisplayName ?? '';
      final cat = item.plug?['plugCategoryIdentifier'] as String? ?? '';
      final element = deriveElement(typeName, cat);
      final classType =
          toClassName(item.classType) ?? _classTypeFromCategory(cat);
      final base = projectBase(item);
      result.add(
        AspectRecord(
          hash: base.hash,
          name: base.name,
          searchName: base.searchName,
          icon: base.icon,
          description: item.displayProperties.description,
          classType: classType,
          element: element,
          fragmentCapacity: _getFragmentCapacity(item),
        ),
      );
    }
    return result;
  }
}

bool _isAspectItem(RawInventoryItem item) {
  final cat = item.plug?['plugCategoryIdentifier'] as String? ?? '';
  final typeName = item.itemTypeDisplayName ?? '';
  return _aspectCategoryRe.hasMatch(cat) || typeName.endsWith('Aspect');
}

int _getFragmentCapacity(RawInventoryItem item) {
  for (final stat in item.investmentStats) {
    if (stat.statTypeHash == _aspectEnergyStatHash) return stat.value;
  }
  return item.socketEntries.length;
}

DestinyClassName? _classTypeFromCategory(String categoryId) {
  final match = RegExp(r'^(titan|hunter|warlock)\.', caseSensitive: false)
      .firstMatch(categoryId);
  if (match == null) return null;
  switch (match.group(1)!.toLowerCase()) {
    case 'titan':
      return DestinyClassName.titan;
    case 'hunter':
      return DestinyClassName.hunter;
    case 'warlock':
      return DestinyClassName.warlock;
    default:
      return null;
  }
}

File: web/NextJS/src/lib/manifest/extractors/aspects.ts
import type { AspectRecord, DestinyClassName } from "../types/records";
import type { Extractor, RawTable, RawTableName } from "../types/services";
import type { RawInventoryItem } from "./rawTypes";
import { iterItems, projectBase, toClassName, deriveElement } from "./common";

const ASPECT_CATEGORY_RE = /^(titan|hunter|warlock|shared)\.(arc|solar|void|stasis|strand|prismatic)\.aspects$/;
const ASPECT_ENERGY_STAT_HASH = 2223994109;

const CLASS_FROM_CAT: Record<string, DestinyClassName> = {
  titan: "Titan",
  hunter: "Hunter",
  warlock: "Warlock",
};

type LoadTable = (table: RawTableName) => Promise<RawTable>;

function isAspectItem(item: RawInventoryItem): boolean {
  const cat = item.plug?.plugCategoryIdentifier ?? "";
  const typeName = item.itemTypeDisplayName ?? "";
  return ASPECT_CATEGORY_RE.test(cat) || typeName.endsWith("Aspect");
}

function getFragmentCapacity(item: RawInventoryItem): number {
  const stat = item.investmentStats?.find((s) => s.statTypeHash === ASPECT_ENERGY_STAT_HASH);
  if (stat !== undefined) return stat.value;
  return item.sockets?.socketEntries?.length ?? 0;
}

function classTypeFromCategory(categoryId: string): DestinyClassName | null {
  const match = /^(titan|hunter|warlock)\./i.exec(categoryId);
  if (!match) return null;
  return CLASS_FROM_CAT[match[1].toLowerCase()] ?? null;
}

async function extractAspects(loadTable: LoadTable): Promise<AspectRecord[]> {
  const itemTable = await loadTable("DestinyInventoryItemDefinition");
  const result: AspectRecord[] = [];

  for (const item of iterItems(itemTable)) {
    if (!isAspectItem(item)) continue;

    const typeName = item.itemTypeDisplayName ?? "";
    const cat = item.plug?.plugCategoryIdentifier ?? "";
    const element = deriveElement(typeName, cat);
    const classType = toClassName(item.classType) ?? classTypeFromCategory(cat);

    result.push({
      ...projectBase(item),
      description: item.displayProperties.description,
      classType,
      element,
      fragmentCapacity: getFragmentCapacity(item),
    });
  }

  return result;
}

export const aspectsExtractor: Extractor<"aspects"> = {
  store: "aspects",
  extract: (loadTable) => extractAspects(loadTable),
};
